lib/TH/generic/THTensorMath.c

```diff
@@ -483,9 +491,7 @@ accreal THTensor_(prodall)(THTensor *tensor)
 void THTensor_(add)(THTensor *r_, THTensor *t, real value)
 {
   THTensor_(resizeAs)(r_, t);
-  if (((THTensor_(isContiguous)(r_) && THTensor_(isContiguous)(t)) ||
-    (THTensor_(isTransposed)(r_) && THTensor_(isTransposed)(t))) &&
-    THTensor_(nElement)(r_) == THTensor_(nElement)(t)) {
```

Comment:
I rolled back these kinds of wrong optimizations.

First I think it is necessary to add more checks into torch.test, since the current test cannot detect these errors.

Then, to apply transposed optimizations. I should separate different situations. For **TH_TENSOR_APPLY2**, there are nine cases for `r_` and `t`--a tensor can be entire transposed, with holes, or contiguous. Thus, nine strategies should be applied. While I think transposed tensors could be handled in the same way as tensors with holes without loss of speed.